pkg/gds/store/index/search.go

```diff
@@ -0,0 +1,169 @@
+func (c *normalizedUnique) PrefixMatch(indexName string, searchPrefixMinLength int) Searcher {
+	return func(query map[string]interface{}) (results []string) {
+		terms, ok := ParseQuery(indexName, query, c.norm)
+		if ok {
+			log.Debug().Str("index", indexName).Strs("terms", terms).Msg("prefix match search")
+			results = make([]string, 0)
+			for _, term := range terms {
+				if result, ok := c.index[term]; ok {
+					// exact match
+					// Only collect unique results in sorted order
+					results = insort(results, result)
+				} else if len(term) >= searchPrefixMinLength {
+					// prefix match
+					for record, result := range c.index {
+						if strings.HasPrefix(record, term) {
+							// Only collect unique results in sorted order
+							results = insort(results, result)
+						}
+					}
+				}
+			}
+		}
+		return results
+	}
+}
+
+// UniqueRecords returns a search function that gathers all of the unique IDs for the
+// terms specified in the search query.
+func (c *normalizedContainer) ContainsRecord(indexName string) Searcher {
+	return func(query map[string]interface{}) (results []string) {
+		terms, ok := ParseQuery(indexName, query, c.norm)
+		if ok {
+			log.Debug().Str("index", indexName).Strs("terms", terms).Msg("contains record search")
+			results = make([]string, 0)
+			for _, term := range terms {
+				if records, ok := c.index[term]; ok {
+					// Add only unique records to the results in sorted order.
+					for _, record := range records {
+						results = insort(results, record)
+					}
+				}
+			}
+		}
+		return results
+	}
+}
+
+// Queries are maps that hold an index name (e.g. "name" or "country") and map it to an
+// indexable key in the index. The query can be either a single string or a list of
+// strings; the parse function extracts the appropriate type and returns a list of
+// strings if the query for the particular index name exists.
+//
+// All values are normalized by the normalize function if provided.
+func ParseQuery(index string, query map[string]interface{}, norm Normalizer) ([]string, bool) {
+	// Check to ensure the index is in the query
+	val, ok := query[index]
+	if !ok {
+		return nil, false
+	}
+
+	// Parse the query based on the query type
+	switch vals := val.(type) {
+	case []string:
+		terms := make([]string, 0, len(vals))
+		for _, term := range vals {
+			if norm != nil {
+				term = norm(term)
+			}
+			terms = insort(terms, term)
+		}
+		return terms, true
+
+	case string:
+		if norm != nil {
+			vals = norm(vals)
+		}
+		return []string{vals}, true
+
+	default:
+		return nil, false
+	}
+}
+
+func insort(arr []string, item string) []string {
+	if len(arr) == 0 {
+		arr = append(arr, item)
+		return arr
+	}
+
+	i := sort.Search(len(arr), func(i int) bool { return arr[i] >= item })
+	if i < len(arr) && arr[i] == item {
+		// value is already in the array
+		return arr
+	}
+
+	arr = append(arr, "")
+	copy(arr[i+1:], arr[i:])
+	arr[i] = item
+	return arr
+}
```

Comment:
This is new - it ensures that search results are returned without duplicates and sorted.

Reply:
Perhaps in a future refactor we can make this a utility function? In the container we have some similar code that's also doing an "insort" type operation.

Reply:
That's a good idea - though a utility function would have to be used with care. Slice operations are notoriously difficult because the underlying array is immutable (in fact, even with this function I went through several revisions with pointer slices and different types of addressing - I opted for likely the least performant version, but the most understandable). 

A slice is really just a low level pointer (really a header with a length, capacity, and pointer to a location in memory). When the slice is changed (e.g. if an append or a copy exceeds the capacity of the underlying array) a new array is allocated and the slice variable is modified. That means that a function cannot modify a slice in place (and you get a lot of weird bugs with empty slices). In this function, by returning the modified slice, we're guaranteeing the caller gets access to it, but at the cost of many escapes (because the array is in the heap, not the stack). 

Another option is to keep a sorted array type around or something like that.